tests/bluetooth/bsim_bt/bsim_test_l2cap/src/main_l2cap_ecred.c

```diff
@@ -0,0 +1,563 @@
+	LOG_DBG("%s", addr);
+
+	SET_FLAG(is_connected);
+}
+
+static void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	LOG_DBG("%s (reason 0x%02x)", addr, reason);
+
+	if (default_conn != conn) {
+		FAIL("Conn mismatch disconnect %s %s)", default_conn, conn);
+		return;
+	}
+
+	bt_conn_unref(default_conn);
+	default_conn = NULL;
+	UNSET_FLAG(is_connected);
+}
+
+BT_CONN_CB_DEFINE(conn_callbacks) = {
+	.connected = connected,
+	.disconnected = disconnected,
+};
+static void send_sdu(int itteration, int chan_idx, int bytes);
+
+void send_sdu_chan_worker(struct k_work *item)
+{
+	struct channel  *ch =
+		CONTAINER_OF(item, struct channel, work);
+
+	send_sdu(ch->itteration, ch->chan_id, ch->bytes_to_send);
+}
+
+static void send_sdu(int itteration, int chan_idx, int bytes)
+{
+	struct bt_l2cap_chan *chan = &channels[chan_idx].le.chan;
+
+	*((int *)channels[chan_idx].payload) = itteration; /*First 4 bytes in sent payload is itteration count*/
+	if (channels[chan_idx].buf != 0) {
+		LOG_ERR("Buf should have been dealocated by now");
+	}
+	struct net_buf *buf = net_buf_alloc(&tx_data_pool, K_NO_WAIT);
+
+	if (buf == 0) {
+		LOG_ERR("Failed to get buff on ch %i, iteration %i should never happen", chan_idx, chan_idx);
+	}
+	channels[chan_idx].buf = buf;
+	net_buf_reserve(buf, BT_L2CAP_CHAN_SEND_RESERVE);
+	net_buf_add_mem(buf, channels[chan_idx].payload, bytes);
+
+	LOG_DBG("bt_l2cap_chan_sending ch: %i bytes: %i itteration: %i", chan_idx, bytes, itteration);
+	int ret = bt_l2cap_chan_send(chan, buf);
+
+	LOG_DBG("bt_l2cap_chan_send returned: %i", ret);
+
+	if (ret < 0) {
+		LOG_DBG("Error: send failed error: %i", ret);
+		channels[chan_idx].buf = 0;
+		net_buf_unref(buf);
+	}
+}
+
+static void test_peripheral_main(void)
+{
+	int err;
+
+	LOG_DBG("*L2CAP ECRED Peripheral started*");
+	init_workqs();
+	err = bt_enable(NULL);
+	if (err) {
+		FAIL("Can't enable Bluetooth (err %d)", err);
+		return;
+	}
+	LOG_DBG("Peripheral Bluetooth initialized.");
+
+	LOG_DBG("Connectable advertising...");
+	err = bt_le_adv_start(BT_LE_ADV_CONN_NAME, ad, ARRAY_SIZE(ad), NULL, 0);
+	if (err) {
+		FAIL("Advertising failed to start (err %d)", err);
+		return;
+	}
+	LOG_DBG("Advertising started.");
+
+	LOG_DBG("Peripheral waiting for connection...");
+	WAIT_FOR_FLAG_SET(is_connected);
+	LOG_DBG("Peripheral Connected.");
+
+	/* Wait a bit to ensure that all LLCP have time to finish */
+	k_sleep(K_MSEC(1000));
+
+	register_l2cap_server();
+
+	connect_num_channels(L2CAP_CHANNELS);
+
+	k_sleep(K_MSEC(500));
+	for (int i = 0; i < SDU_SEND_COUNT; i++) {
+		channels[0].itteration = i;
+		channels[0].bytes_to_send = DATA_MTU - 500;
+		k_work_submit_to_queue(&my_work_q0, &channels[0].work);
+		channels[1].itteration = i;
+		channels[1].bytes_to_send = DATA_MPS - 2;
+		k_work_submit_to_queue(&my_work_q1, &channels[1].work);
+		k_sleep(K_MSEC(5000));
```

Comment:
Why are we waiting 5 seconds between each TX? Could we instead just send whenever the previous TX is done using the `sent` callback?

Reply:
Can you first perhaps elaborate on why we are waiting 5 seconds? 

If it is to make sure the previous data has been sent, then there's a `sent` callback for L2CAP that can be used to determine that, and thus avoiding any `k_sleep` with an absolute value.